web-prover-core/src/parser/config.rs

```diff
@@ -221,4 +248,179 @@ mod tests {
+            id: "link_href".to_string(),
+            description: "Link URL".to_string(),
+            selector: vec![".link".to_string()],
+            attribute: Some("href".to_string()),
+            extractor_type: ExtractorType::String
+        ),
+      ],
+    };
+
+    let result = config.extract_and_validate(&html).unwrap();
+
+    assert_eq!(result.errors.len(), 0);
+    assert_eq!(result.values.len(), 5);
+
+    assert_eq!(result.values["username"], json!("John Doe"));
+    assert_eq!(result.values["age"], json!("30"));
+    assert_eq!(result.values["active"], json!("true"));
+    let tags = result.values["tags"].as_array().unwrap();
+    assert!(tags.contains(&json!("developer")));
+    assert!(tags.contains(&json!("rust")));
+    assert!(tags.contains(&json!("web")));
+    assert_eq!(result.values["link_href"], json!("https://example.com"));
+  }
+
+  #[test]
+  fn test_extract_and_validate_html_with_failures() {
+    let html = r#"
+      <!DOCTYPE html>
+      <html>
+      <body>
+        <div class="user-info">
+          <h1 id="username">John Doe</h1>
+        </div>
+      </body>
+      </html>
+    "#
+    .to_string()
+    .into_bytes();
+
+    let config = ExtractorConfig {
+      format:     DataFormat::Html,
+      extractors: vec![
+        extractor!(
+            id: "username".to_string(),
+            description: "User's name".to_string(),
+            selector: vec!["#username".to_string()],
+            extractor_type: ExtractorType::String
+        ),
+        extractor!(
+            id: "missing_element".to_string(),
+            description: "Missing element".to_string(),
+            selector: vec![".non-existent".to_string()],
+            extractor_type: ExtractorType::String
+        ),
+        extractor!(
+            id: "optional_missing".to_string(),
+            description: "Optional missing element".to_string(),
+            selector: vec![".optional".to_string()],
+            extractor_type: ExtractorType::String,
+            required: false
+        ),
+      ],
+    };
+
+    let result = config.extract_and_validate(&html).unwrap();
+
+    assert_eq!(result.errors.len(), 1);
+    assert_eq!(result.values.len(), 1);
+
+    assert_eq!(result.values["username"], json!("John Doe"));
+    assert!(result.errors[0].contains("No elements found for selector"));
+  }
+
+  #[test]
+  fn test_extract_and_validate_html_with_invalid_input() {
+    let config = ExtractorConfig {
+      format:     DataFormat::Html,
+      extractors: vec![extractor! {
+        id:             "username".to_string(),
+        description:    "User's name".to_string(),
+        selector:       vec!["#username".to_string()],
+        extractor_type: ExtractorType::String,
+      }],
+    };
+
+    // Test with non-UTF-8 bytes
+    let bad_input = b"\xFF\xFEInvalid UTF-8".to_vec();
+    let result = config.extract_and_validate(&bad_input);
+    assert!(matches!(result, Err(ExtractorError::InvalidHtml(_))));
+  }
+
+  #[test]
```

Comment:
I really appreciate this test driven development that you do. It makes it really easy to see how what you've added works and what it handles. Thanks!